feat: succesfully complete all challenges

Menendez2004 opened the pull request:
Completed all assigned challenges, addressing the required tasks. All solutions are working

Menendez2004 committed:
feat: succesfully complete all challenges
Completed all assigned challenges, addressing the required tasks. All solutions are working

Menendez2004 committed:
feat: ERD challenge
ERD design already done for the laast challenge

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
 -- Your answers here:
 -- 1
+SELECT
+    type,
+    SUM(mount) AS total_money
+FROM
+    accounts
+GROUP BY
+    type
+ORDER BY
+    type ASC;
```
Good solution here.

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+            user_id
+        HAVING
+            COUNT(*) >= 2
+    );
```
You can separate the sub-query with `WITH query_name AS` syntax to improve readability. 
With the current implementation, you should add an alias because every derived table or subquery in PostgreSQL must have an alias to ensure proper functionality and readability. 
For reference. 4.2.1.3. Subqueries: https://www.postgresql.org/docs/7.3/queries-table-expressions.html

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+
+
 
 -- 5
```
In this exercise at the begging said `In this part, you need to CRATE A TRANSACTION with the following steps`, so that means all the following items should be within a transaction. With that in mind the solution for `e` and `f` literals will change.

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+ORDER BY
+    total_money DESC
+LIMIT
+    3;
```
I think here you misunderstood the exercise because you have to identify the type of movements that every account has based on that sum or subtract from the account balance and then when you have that result query the top 3 accounts.

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+            WHEN m.account_to = a.id THEN m.mount  
+        END
+    )) AS final_account_amount,
+    ARRAY_AGG(ROW(
+        m.type, 
+        m.account_from, 
+        m.account_to, 
+        m.mount
+    )) AS movements
```
I think this is not needed as the exercise does not mention including it.

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+--(B) Add a new movement with the information: from: 3b79e403-c788-495a-a8ca-86ad7643afaf make a transfer to fd244313-36e5-4a17-a27c-f8265bc46590 mount: 50.75
+DO $$ 
+DECLARE
+    first_account UUID := '3b79e403-c788-495a-a8ca-86ad7643afaf';
+    second_account UUID := 'fd244313-36e5-4a17-a27c-f8265bc46590';
```
You can improve the name of the variables to be more descriptive.

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+DECLARE
+    first_account UUID := '3b79e403-c788-495a-a8ca-86ad7643afaf';
+    second_account UUID := 'fd244313-36e5-4a17-a27c-f8265bc46590';
+    transfer_amount NUMERIC := 80;
```
The transfer amount is `50.75`

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+-----------------------------------------
+--Shows the result of the tranfer--
+-----------------------------------------
+SELECT 
+    m.type, 
+    m.account_from, 
+    m.account_to, 
+    m.mount AS transferred_amount
+FROM movements m
+WHERE m.account_from = '3b79e403-c788-495a-a8ca-86ad7643afaf'
+   OR m.account_to = '3b79e403-c788-495a-a8ca-86ad7643afaf'
+ORDER BY m.created_at ASC;
```
Nice you included this, that helped me when I was reviewing!

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+    WHERE a.id = first_account;
+
+    IF current_balance < out_amount THEN
+        RAISE EXCEPTION 'Saldo insuficiente para realizar el movimiento, transacción rechazada';
```
The messages should be in english

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+	
+EXCEPTION
+WHEN OTHERS THEN
+    RAISE NOTICE '*** Error procesando la transacción con monto %. Saldo disponible: % % ** Haciendo ROLLBACK **', 
```
Same comment here. English message

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+
+--(d). Put your answer here if the transaction fails(YES/NO): Yes  the  is failed cuz the mount is higher than the current balance 
+
+--(e and f). If the transaction fails, make the correction on step c to avoid the failure (the way to avoid the failure is tranfer less money than the current balance)
```
The solution will change based on my comment to wrap the whole exercise in a transaction.

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+SELECT 
+    a.id AS account_id,
+    a.mount + 
+	SUM(
+        CASE 
+            WHEN m.account_from = a.id THEN -m.mount 
+            WHEN m.account_to = a.id THEN m.mount  
+        END
+    )AS total_money
```
You have to do the same query several times. Remember do not repeat yourself.

fabio248 commented on src/answers.sql:
```diff
@@ -1,15 +1,290 @@
+    u.email = 'Kaden.Gusikowski@gmail.com'
+ORDER BY 
+    a.type ASC,
+    m.created_at ASC;
```
Nice solution in the last 3 exercises.

fabio248 reviewed (commented):
I left you some comments on improvements and corrections that you can make.
We will talk about the ERD in the feedback meeting.

Menendez2004 committed:
fix: all issues was solved
-problems with the comments lenguage
-same query several times
-confusion with the exercise #4
